src/java/org/apache/cassandra/service/MigrationManager.java

```diff
@@ -149,25 +157,50 @@ public void run()
                                 endpoint,
                                 theirVersion,
                                 currentVersion);
+                        removeEndpointFromSchemaPullVersion(theirVersion, endpoint);
                         return;
                     }
 
                     if (Schema.instance.getVersion().equals(currentVersion))
                     {
                         logger.debug("not submitting migration task for {} because our versions match", endpoint);
+                        removeEndpointFromSchemaPullVersion(theirVersion, endpoint);
                         return;
                     }
-                    logger.debug("submitting migration task for endpoint {}, endpoint schema version {}, and our schema version",
-                            endpoint,
-                            currentVersion,
-                            Schema.instance.getVersion());
-                    submitMigrationTask(endpoint);
+                    logger.debug("submitting migration task for endpoint {}, endpoint schema version {}, and our schema version {}",
+                            SafeArg.of("endpoint", endpoint),
+                            SafeArg.of("endpointVersion", currentVersion),
+                            SafeArg.of("schemaVersion", Schema.instance.getVersion()));
+                    submitMigrationTask(endpoint, currentVersion);
                 }
             };
+            addEndointToSchemaPullVersion(theirVersion, endpoint);
             ScheduledExecutors.nonPeriodicTasks.schedule(runnable, MIGRATION_DELAY_IN_MS, TimeUnit.MILLISECONDS);
         }
     }
 
+    public static void addEndointToSchemaPullVersion(UUID version, InetAddress endpoint) {
+        scheduledSchemaPulls.putIfAbsent(version, new HashSet<>());
+        scheduledSchemaPulls.computeIfPresent(version, (v, s) -> {
+            s.add(endpoint);
+            return s;
```

Comment:
do we have a race between these two calls to the map and the endpoint getting removed?

also nit typo in "endoint"

Reply:
Updated!